control/diagnostics_tool/src/system_state.c

```diff
@@ -0,0 +1,264 @@
+
+/* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES */
+/* OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND */
+/* NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT */
+/* HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, */
+/* WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING */
+/* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR */
+/* OTHER DEALINGS IN THE SOFTWARE. */
+/************************************************************************/
+
+
+/**
+* @file system_state.c
+* @brief System State Source.
+*
+*/
+
+
+
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "macros.h"
+#include "can_monitor.h"
+#include "terminal_print.h"
+#include "system_state.h"
+
+
+
+
+// *****************************************************
+// static global data
+// *****************************************************
+
+
+static current_system_state_s system_state;
+
+
+
+
+// *****************************************************
+// static definitions
+// *****************************************************
+
+
+//
+static int update_steering_state()
+{
+    int ret = NOERR;
+
+    const can_frame_s * const steering_command_frame =
+            get_can_msg_array_index_reference( PS_CTRL_MSG_ID_STEERING_COMMAND );
+
+    const can_frame_s * const steering_report_frame =
+            get_can_msg_array_index_reference( PS_CTRL_MSG_ID_STEERING_REPORT );
+
+    if( steering_command_frame == NULL  ||  steering_report_frame == NULL )
+    {
+        ret = ERROR;
+    }
+
+    if( ret == NOERR )
+    {
+        ret = analyze_steering_state(
+                &system_state.steering_module_state,
+                steering_command_frame,
+                steering_report_frame );
+    }
+
+    return ret;
+}
+
+
+//
+static int update_throttle_state()
+{
+    int ret = NOERR;
+
+    const can_frame_s * const throttle_command_frame =
+            get_can_msg_array_index_reference( PS_CTRL_THROTTLE_COMMAND_ID );
+
+    const can_frame_s * const throttle_report_frame =
+            get_can_msg_array_index_reference( PS_CTRL_MSG_ID_THROTTLE_REPORT );
+
+    if( throttle_command_frame == NULL || throttle_report_frame == NULL )
+    {
+        ret = ERROR;
+    }
+
+    if( ret == NOERR )
+    {
+        ret = analyze_throttle_state(
+                &system_state.throttle_module_state,
+                throttle_command_frame,
+                throttle_report_frame );
+    }
+
+    return ret;
+}
+
+
+//
+static int update_brake_state()
+{
+    int ret = NOERR;
+
+    const can_frame_s * const brake_command_frame =
+            get_can_msg_array_index_reference( PS_CTRL_MSG_ID_BRAKE_COMMAND );
+
+    const can_frame_s * const brake_report_frame =
+            get_can_msg_array_index_reference( PS_CTRL_MSG_ID_BRAKE_REPORT );
+
+    if( brake_command_frame == NULL  ||  brake_report_frame == NULL )
```

Comment:
Same as above.  In this case, it's best to not assume that everything is going to work :)